oak_restricted_kernel/src/interface.rs

```diff
@@ -0,0 +1,94 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Exported interface for interacting with the kernel from C.
+
+use alloc::alloc::{alloc, dealloc, Layout};
+use core::{
+    cell::{LazyCell, RefCell},
+    ffi::c_void,
+    ptr::null_mut,
+};
+use hashbrown::HashMap;
+use spinning_top::{const_spinlock, Spinlock};
+use static_assertions::const_assert;
+
+/// The maximum alignment that we need to support any C type that has fundamental alignment.
+const MAX_ALIGNMENT: usize = core::mem::align_of::<libc::max_align_t>();
```

Comment:
On x86_64 this is 16 bytes (the size of a 128-bit float), but it is architecture-dependent. If the allocation size is smaller than 16 bytes the alignment is calculated as the next power of two that is larger or equal to the requested allocation size.

Reply:
Why is this alignment necessary?

Reply:
This is required for interoperability with C. The C compiler will make assumptions about the alignment of structs in memory and optimise the code based on those assumptions. If the alignments do not match these assumptions it can lead to undefined behaviour.